### PR TITLE
Required keyword on various types

### PR DESCRIPTION
**Pull request by JR-Morgan**

### CODE REVIEW

**Comment by JR-Morgan** [comment]
Ready for review, but don't merge until I say so!

File: src/Speckle.Objects/Geometry/Polyline.cs [hunk 1]
```diff
@@ -12,170 +10,29 @@ namespace Speckle.Objects.Geometry;
 /// A polyline curve, defined by a set of vertices.
 /// </summary>
 [SpeckleType("Objects.Geometry.Polyline")]
-public class Polyline : Base, ICurve, IHasArea, IHasBoundingBox, IConvertible, ITransformable
```
Comment (AlanRynne) [inline src/Speckle.Objects/Geometry/Polyline.cs:15]: I approve of removing I convertible, but I'm fairly certain this was being done for to allow a Polyline to be casted to a PolyCurve in some cases?

Obviously this was not the right way to do this either, so the removal is good, but if this is needed maybe we should implement explicit casting instead.

File: src/Speckle.Objects/Geometry/Polyline.cs [hunk 2]
```diff
@@ -12,170 +10,29 @@ namespace Speckle.Objects.Geometry;
...
-  [JsonIgnore, Obsolete("Use " + nameof(GetPoints) + " Instead", true)]
-  public List<Point> points => GetPoints();
-
-  /// <inheritdoc/>
-  public object ToType(Type conversionType, IFormatProvider provider)
-  {
-    if (conversionType == typeof(Polycurve))
-    {
-      return (Polycurve)this;
-    }
```
Comment (AlanRynne) [inline src/Speckle.Objects/Geometry/Polyline.cs:74]: I'm not sure how this even worked... but OK 😅

File: src/Speckle.Objects/Other/Instance.cs [hunk 3]
```diff
@@ -149,13 +149,14 @@ protected override IEnumerable<Base> GetTransformableGeometry()
   public Plane GetInsertionPlane()
   {
     // TODO: UPDATE!
```
Comment (AlanRynne) [inline src/Speckle.Objects/Other/Instance.cs:151]: This vague TODO is a bit scary... Are we missing anything critical that is yet to be implemented here?

I'm aware this is unrelated to this PR. Just asking

File: src/Speckle.Objects/Structural/CSI/Geometry/CSINode.cs [hunk 4]
```diff
@@ -40,7 +41,14 @@ public CSINode(
...
         AxisType.Cartesian,
-        new Plane(new Point(0, 0), new Vector(0, 0, 1), new Vector(1, 0, 0), new Vector(0, 1, 0))
+        new Plane
+        {
+          origin = new Point(0, 0),
+          normal = new Vector(0, 0, 1),
+          xdir = new Vector(1, 0, 0),
+          ydir = new Vector(0, 1, 0),
+          units = Units.Meters, //Not sure if defaulting to meters is correct, but it was what we were doing previously inside Plane's ctor
+        }
```
Comment (AlanRynne) [inline src/Speckle.Objects/Structural/CSI/Geometry/CSINode.cs:51]: It is most likely not. TBH I'm leaning towards nuking all structural objects (at least the app specific ones) to be able to start anew and not have to drag these refactors into classes we're most likely not keeping anyway.

**Review by AlanRynne — COMMENTED** [review]
Not requesting changes but we may end up opening some tickets around some of my comments

File: src/Speckle.Objects/Geometry/Brep.cs [hunk 5]
```diff
@@ -565,21 +573,22 @@ internal void OnDeserialized(StreamingContext context)
...
       var e = Edges[i];
-      lock (e)
+      var existing = e;
+      lock (existing)
       {
         if (e.Brep != null)
         {
-#pragma warning disable CS0728 // Possibly incorrect assignment to local which is the argument to a using or lock statement
-          e = new BrepEdge(
-#pragma warning restore CS0728 // Possibly incorrect assignment to local which is the argument to a using or lock statement
```
[the remark this answers is not in the document]
Comment (JR-Morgan) [reply src/Speckle.Objects/Geometry/Brep.cs:574]: I assigned a new variable a couple lines up, `existingBrepEdge`. Locking works on objects, not variables, this warning reminds you about that when you're reassigning the variable.

File: src/Speckle.Objects/Geometry/Brep.cs [hunk 6]
```diff
@@ -69,15 +36,11 @@ public List<double> SurfacesValue
   {
     get
     {
-      var list = new List<double>();
-      if (Surfaces != null)
+      var list = new List<double>(Surfaces.Count);
```
[the remark this answers is not in the document]
Comment (JR-Morgan) [reply src/Speckle.Objects/Geometry/Brep.cs:39]: oh damn, you're absolutely right,
I'll get rid of it, as its misleading

File: src/Speckle.Objects/Geometry/Brep.cs [hunk 7]
```diff
@@ -169,7 +132,7 @@ public List<double> VerticesValue
   {
     get
     {
-      var list = new List<double>();
+      var list = new List<double>(Vertices.Count + 1);
```
[the remark this answers is not in the document]
Comment (JR-Morgan) [reply src/Speckle.Objects/Geometry/Brep.cs:135]: Good catch, clearly should be `vertices.Count * 3` and +1 for the unit encoding

File: src/Speckle.Objects/Geometry/Brep.cs [hunk 8]
```diff
@@ -356,7 +339,8 @@ public List<int> TrimsValue
           CurveIndex = value[i + 5],
           IsoStatus = value[i + 6],
           TrimType = (BrepTrimType)value[i + 7],
-          IsReversed = value[i + 8] == 1
+          IsReversed = value[i + 8] == 1,
+          Domain = Interval.UnitInterval, //TODO: This is a problem, see CXPLA-28
```
[the remark this answers is not in the document]
Comment (JR-Morgan) [reply src/Speckle.Objects/Geometry/Brep.cs:343]: done

File: src/Speckle.Objects/Geometry/Brep.cs [hunk 9]
```diff
@@ -320,7 +302,7 @@ public List<int> TrimsValue
   {
     get
     {
-      List<int> list = new();
+      List<int> list = new(Trims.Count * 9);
```
[the remark this answers is not in the document]
Comment (JR-Morgan) [reply src/Speckle.Objects/Geometry/Brep.cs]: Added a const

File: src/Speckle.Objects/Geometry/Polyline.cs (hunk 1 above)
Comment (JR-Morgan) [reply src/Speckle.Objects/Geometry/Polyline.cs:15]: I don't think we were ever using the IConvertible ness...

You would have had to either manually call ToType<Polyline>(), or use the `[ChangeType](https://learn.microsoft.com/en-us/dotnet/api/system.convert.changetype?view=net-8.0#system-convert-changetype(system-object-system-type)) helper function...

I don't think casts are appropriate either, but I would accept a `ToPolyLine` function or similar... if we need it that is.

File: src/Speckle.Objects/Other/Instance.cs (hunk 3 above)
Comment (JR-Morgan) [reply src/Speckle.Objects/Other/Instance.cs:151]: image

These are some very old comments from @clairekuang . I suspect they are not a problem.
But @clairekuang if you don't mind having a quick look in-case there is indeed some gaps that need filling. (low priority)

File: src/Speckle.Objects/Structural/CSI/Geometry/CSINode.cs (hunk 4 above)
Comment (JR-Morgan) [reply src/Speckle.Objects/Structural/CSI/Geometry/CSINode.cs:51]: Lets chat sometime, 
we'll need to nuke some eventually anyway, but I'm a little hesitant to start lighting the dynamite just yet.

**Review by AlanRynne — APPROVED** [review]

File: src/Speckle.Objects/Geometry/Circle.cs [hunk 10]
```diff
@@ -67,9 +51,9 @@ public List<double> ToList()
   {
     var list = new List<double>();
 
-    list.Add(radius ?? 0);
-    list.Add(domain?.start ?? 0);
-    list.Add(domain?.end ?? 1);
+    list.Add(radius);
```
Comment (clairekuang) [inline src/Speckle.Objects/Geometry/Circle.cs:54]: this is the only change i'm not 100% sure about, will check some circle curves in breps for deserialization issues

File: src/Speckle.Objects/Geometry/Ellipse.cs [hunk 11]
```diff
@@ -93,10 +48,10 @@ public Ellipse(
   public List<double> ToList()
   {
     var list = new List<double>();
-    list.Add(firstRadius ?? 0);
-    list.Add(secondRadius ?? 0);
-    list.Add(domain?.start ?? 0);
-    list.Add(domain?.end ?? 0);
+    list.Add(firstRadius);
```
Comment (clairekuang) [inline src/Speckle.Objects/Geometry/Ellipse.cs:51]: I guess same here for ellipses in breps

File: src/Speckle.Objects/Other/Instance.cs (hunk 3 above)
Comment (clairekuang) [reply src/Speckle.Objects/Other/Instance.cs:151]: I'm ok removing the comments, if any bugs are reported with non-0,0,0 based base points, we can address them at that point

File: src/Speckle.Objects/Other/Instance.cs [hunk 12]
```diff
@@ -149,13 +149,14 @@ protected override IEnumerable<Base> GetTransformableGeometry()
...
-    var plane = new Plane(
-      typedDefinition.basePoint ?? new Point(0, 0, 0, units),
-      new Vector(0, 0, 1, units),
-      new Vector(1, 0, 0, units),
-      new Vector(0, 1, 0, units),
-      units
-    );
+    var plane = new Plane()
+    {
+      origin = typedDefinition.basePoint ?? new Point(0, 0, 0, units),
```
Comment (clairekuang) [inline src/Speckle.Objects/Other/Instance.cs:154]: why doesn't this `Point` use the new construction?

File: src/Speckle.Objects/Other/Revit/RevitInstance.cs [hunk 13]
```diff
@@ -56,13 +56,14 @@ public override IEnumerable<ITransformable> GetTransformedGeometry()
...
     // TODO: Check for Revit in GH/DYN
-    var plane = new Plane(
-      new Point(0, 0, 0, units),
-      new Vector(0, 0, 1, units),
-      new Vector(1, 0, 0, units),
-      new Vector(0, 1, 0, units),
-      units
-    );
+    var plane = new Plane()
+    {
```
Comment (clairekuang) [inline src/Speckle.Objects/Other/Revit/RevitInstance.cs:60]: same comment, do these new points and vectors need to use the new construction?

File: src/Speckle.Objects/Other/Instance.cs (hunk 12 above)
Comment (clairekuang) [reply src/Speckle.Objects/Other/Instance.cs:154]: nvm, point and vector haven't been updated to use `required` keyword

**Review by clairekuang — APPROVED** [review]

File: src/Speckle.Objects/Geometry/Circle.cs (hunk 10 above)
Comment (JR-Morgan) [reply src/Speckle.Objects/Geometry/Circle.cs:54]: so any `null`s in json will become `0` post this change... this is only a problem for the `end` part of the domain, since sometimes, like here, we're defaulting to `1`

so the deserailizer won't throw, but conversions might.
I looked through the latest v2 codebase, and the current v3 codebase, and I can't see anywhere that might potentially leave a null end point... perhaps some historical data might, but I think we'll have to live with some loss v2 -> v3
